### ISSUE TITLE
Incomplete Intellisense/Autocomplete in Generic Func

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#40337]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/836185/incomplete-intellisenseautocomplete-in-generic-fun.html)._

---
Given the following C# code:

```C#
using System;

public class C
{
    public void Main()
    {
        var x = new FakeModelB();
        Foo(x, model => model. StrB, "");
    }

    public void Foo<ModelType, PropertyType>(ModelType model, Func<ModelType, PropertyType> func, PropertyType val)
    { }
}

public class FakeModelB
{
    public string StrB { get; set; }
}
```

When writing the line to call 'Foo', Intellisense is unable to provide correct AutoComplete information for the properties of 'model'.

If you remove the third parameter (val) from the signature of 'Foo', then Intellisense works correctly, as is able to suggest 'StrB'. But with the third parameter in place, Intellisense acts as if 'model' is of type 'object'.

Hovering over 'model' shows a tooltip which prooves that Studio knows it is of type 'FakeModelB', it just doesn't give the correct AutoComplete list.

---
#### Visual Studio Feedback System on 11/27/2019, 10:57 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by dpoeschl** [comment]
Regression from VS 2017.

**Comment by dpoeschl** [comment]
Here's another simple repro. Expecting `Length` to be offered at the $$.

```C#
using System;

public class C
{
    public void Test()
    {
        Foo("", s => Console.WriteLine(s.$$));
    }

    public void Foo<T>(T t, Action<T> a) { }
}
```

**Comment by dpoeschl** [comment]
Here's another very similar repro, this time using Parallel.ForEach:

```C#
using System;
using System.Threading.Tasks;
using System.Collections.Generic;

class C
{
    void M()
    {

        IEnumerable<string> strings = null;
        Parallel.ForEach(strings, s => Console.WriteLine(s.Length));
    }

    // Signature of Parallel.ForEach:
    // public static ParallelLoopResult ForEach<TSource>(IEnumerable<TSource> source, Action<TSource> body);
}
```

**Comment by sujayvsarma** [comment]
Thank you to @dpoeschl for porting the ticket over.

**Comment by JoeRobich** [comment]
This issue affects Unity users of the VS Code C# extension. See https://github.com/OmniSharp/omnisharp-vscode/issues/3469

**Comment by jprob** [comment]
Is there a workaround until this is fixed?

**Comment by stevenxi** [comment]
Is there ETA on this bug fix?

It also impacts .Join(), which is quite annoying.  Although not a blocker.

![Capture](https://user-images.githubusercontent.com/3340119/83576059-c88a4080-a528-11ea-8a20-2b21379805a1.JPG)
![Capture2](https://user-images.githubusercontent.com/3340119/83576060-c922d700-a528-11ea-9745-250b55e2f649.JPG)

**Comment by CyrusNajmabadi** [comment]
I think i fixed this.  Validating now.

**Comment by CyrusNajmabadi** [comment]
Yup fixed with https://github.com/dotnet/roslyn/pull/44601

![image](https://user-images.githubusercontent.com/4564579/83576627-ff3f6700-a4e6-11ea-8fb0-d4f00a20a4a2.png)

**Comment by stevenxi** [comment]
Hi @CyrusNajmabadi 

Thanks for fixing this.  Do you have any idea if this fix can be available as part of 16.6.x release?

**Comment by jmarolf** [comment]
This will be in the 16.7 release of VS
